core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,153 @@
+import static io.trino.sql.planner.plan.Patterns.mergeProcessor;
+import static io.trino.sql.planner.plan.Patterns.mergeWriter;
+import static io.trino.sql.planner.plan.Patterns.project;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static io.trino.sql.planner.plan.Patterns.tableFinish;
+import static io.trino.sql.planner.plan.Patterns.tableScan;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This version support only constant updates
+ * and fall back to default behaviour in all other cases
+ */
+public class PushMergeWriterUpdateIntoConnector
+        implements Rule<TableFinishNode>
+{
+    private static final Capture<MergeWriterNode> MERGE_WRITER_NODE_CAPTURE = newCapture();
+    private static final Capture<MergeProcessorNode> MERGE_PROCESSOR_NODE_CAPTURE = newCapture();
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+    private static final Capture<ProjectNode> PROJECT_NODE_CAPTURE = newCapture();
+
+    private static final Pattern<TableFinishNode> PATTERN =
+            tableFinish().with(source().matching(
+                    mergeWriter().capturedAs(MERGE_WRITER_NODE_CAPTURE).with(source().matching(
+                            mergeProcessor().capturedAs(MERGE_PROCESSOR_NODE_CAPTURE).with(source().matching(
+                                    project().capturedAs(PROJECT_NODE_CAPTURE).with(source().matching(
+                                            tableScan().capturedAs(TABLE_SCAN)))))))));
+
+    public PushMergeWriterUpdateIntoConnector(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer, Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    private final Metadata metadata;
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<TableFinishNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TableFinishNode node, Captures captures, Context context)
+    {
+        MergeWriterNode mergeWriter = captures.get(MERGE_WRITER_NODE_CAPTURE);
+        MergeProcessorNode mergeProcessor = captures.get(MERGE_PROCESSOR_NODE_CAPTURE);
+        ProjectNode project = captures.get(PROJECT_NODE_CAPTURE);
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        TableHandle tableHandle = mergeWriter.getTarget().getHandle();
+        Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(context.getSession(), tableHandle);
+
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(mergeProcessor.getMergeRowSymbol()).getChildren();
```

Comment:
add `checkState(orderedColumnNames.size() == mergeAssignments.size(), "...)`

Reply:
mergeAssignments contains also other things : 

```
.addAll(columnSymbols)
                .add(operationSymbol)
                .add(rowIdSymbol)
                .add(insertFromUpdateSymbol)
```

So should we do such check : 

`checkState(orderedColumnNames.size() == mergeAssignments.size() - 3, "...)` ?

Reply:
Can filter these additional symbols then? What guarantees the order of the mergeAssignments?

Reply:
So it always `// All data columns in table order` plus 3 additional fields.

Reply:
Ok please add this reasoning to the code as a comment and place `checkState(orderedColumnNames.size() == mergeAssignments.size() - 3, "...)`

Reply:
refactored